File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.cpp
class Solution {
public:

    void print2d(vector<vector<int>>& v){
        for(auto &x: v){
            for(auto &y: x){
                cout << y << " ";
            }
            cout << endl;
        }
    }

    vector<vector<int>> vis, ans, pf, at;
    int rows, cols; 

    void dfs(vector<vector<int>>& heights, vector<vector<int>>& visited, int row, int col, int prev){
        if(row < 0 || row >= rows || col < 0 || col >= cols){
            return;
        }

        if(visited[row][col] != -1){
            return;
        }

        // now check if the current value is higher than parent or not, only then we can flow
        int val = heights[row][col];
        if(val < prev){
            return;
        }

        visited[row][col] = 1;

        // now continue with dfs
        dfs(heights, visited, row + 1, col, val);
        dfs(heights, visited, row - 1, col, val);
        dfs(heights, visited, row, col + 1, val);
        dfs(heights, visited, row, col - 1, val);
    }

    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        rows = heights.size();
        cols = heights[0].size();

        pf.resize(rows, vector<int>(cols, -1));
        at.resize(rows, vector<int>(cols, -1));

        for(int i = 0; i < rows; i++){
            // both atlantic and pacific
            dfs(heights, pf, i, 0, heights[i][0]);
            dfs(heights, at, i, cols - 1, heights[i][cols - 1]);
        }

        for(int i = 0; i < cols; i++){
            // both atlantic and pacific
            dfs(heights, pf, 0, i, heights[0][i]);
            dfs(heights, at, rows - 1, i, heights[rows -1][i]);
        }

        // print2d(pf);
        // cout << endl;
        // print2d(at);

        for(int i = 0; i < rows; i++){
            for(int j = 0; j < cols; j++){
                if(at[i][j] == 1 && pf[i][j] == 1){
                    ans.push_back({i,j});
                }
            }
        }

        return ans; 
    }
};

File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.py
drs = [(0, 1), (0, -1), (1, 0), (-1, 0)]

def dfs(vis, h, i, j): 
    if vis[i][j] == 1:
        return

    vis[i][j] = 1
    # each direction
    for d in drs: 
        nr, nc = i + d[0], j + d[1]
        # check if valid 
        if (nr >= 0 and nr < rows and nc >= 0 and nc < cols):
            # valid
            if h[nr][nc] >= h[i][j] and vis[nr][nc] == 0: 
                dfs(vis, h, nr, nc)


class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        # boundaries to inside increasing dfs
        global rows, cols
        rows, cols = len(heights), len(heights[0])
        
        # visited
        pcf = [[0]*cols for _ in range(rows)]
        atl = [[0]*cols for _ in range(rows)]

        # pcf = row 0
        for i in range(cols): 
            if pcf[0][i] == 0: 
                dfs(pcf, heights, 0, i)
        # pcf = col 0
        for i in range(rows):
            if pcf[i][0] == 0: 
                dfs(pcf, heights, i, 0)

        # atl = col n - 1
        for i in range(rows): 
            if atl[i][cols - 1] == 0: 
                dfs(atl, heights, i, cols - 1)
        
        # atl = row n - 1
        for i in range(cols):
            if atl[rows - 1][i] == 0: 
                dfs(atl, heights, rows - 1, i)

        # print(pcf)
        # print(atl)

        ans = []
        for i in range(rows):
            for j in range(cols):
                if atl[i][j] == 1 and pcf[i][j] == 1:
                    ans.append([i,j])

        return ans




# Synced seamlessly with LeetHub Pro
# Pro features: https://bit.ly/leethubpro | Free version: https://bit.ly/leethubv4
# Get it here: https://chromewebstore.google.com/detail/bcilpkkbokcopmabingnndookdogmbna